engine/docker/docker.go

```diff
@@ -525,3 +537,33 @@ func (e *Docker) setProxyInDockerDaemon(ctx context.Context, pipelineConfig *spe
 		}
 	}
 }
+
+// SoftStop stops the container giving them a 30 seconds grace period. The signal sent by ContainerStop is SIGTERM.
+// After the grace period, the container is killed with SIGKILL.
+// After all the containers are stopped, they are removed only when the status is not "running" or "removing".
+func (e *Docker) SoftStop(ctx context.Context, name string) {
+	timeout := 30 * time.Second
+	if err := e.client.ContainerStop(ctx, name, &timeout); err != nil {
+		logrus.WithField("container", name).WithField("error", err).Warnln("failed to stop the container")
+	}
+
+	// Before removing the container we want to be sure that it's in a healthy state to be removed.
+
+	containerStatus, err := e.client.ContainerInspect(ctx, name)
+
+	if err != nil {
+		logrus.WithField("container", name).WithField("error", err).Warnln("failed to retrieve container stats")
+	}
+
+	for {
```

Comment:
Just curious - is there anything preventing us from going in an infinite loop here in case `ContainerInspect` returns an error? Even though it should not happen, there should be some max retries.
Couldn't we force kill in case the `ContainerStop` happens successfully (would it exit out only after 30 seconds?). IMO if the container has been given 30 seconds to terminate gracefully, that should be enough time and we could terminate forcefully afterwards

Reply:
Yeah makes sense to add something to be sure that even in the unlikely scenario of the 30seconds passing if we are still trying to do stuff, just break from the loop and continue